controller/appcontroller.go

```diff
@@ -272,9 +272,35 @@ func NewApplicationController(
 					return fmt.Errorf("application controller deployment replicas is not set or is less than 0, replicas: %d", appControllerDeployment.Spec.Replicas)
 				}
 				shard := env.ParseNumFromEnv(common.EnvControllerShard, -1, -math.MaxInt32, math.MaxInt32)
-				if _, err := sharding.GetOrUpdateShardFromConfigMap(kubeClientset.(*kubernetes.Clientset), settingsMgr, int(*appControllerDeployment.Spec.Replicas), shard); err != nil {
+				shard, err := sharding.GetOrUpdateShardFromConfigMap(kubeClientset.(*kubernetes.Clientset), settingsMgr, int(*appControllerDeployment.Spec.Replicas), shard)
+				if err != nil {
 					return fmt.Errorf("error while updating the heartbeat for to the Shard Mapping ConfigMap: %w", err)
 				}
+
+				// update the shard number in the clusterSharding, and resync all applications if the shard number is updated
+				if ctrl.clusterSharding.UpdateShard(shard) {
+					// update shard number in stateCache
+					ctrl.stateCache.UpdateShard(shard)
+
+					// resync all applications
+					apps, err := ctrl.appLister.List(labels.Everything())
+					if err != nil {
+						return err
+					}
+					for _, app := range apps {
+						if !ctrl.canProcessApp(app) {
+							return nil
+						}
+						key, err := cache.MetaNamespaceKeyFunc(app)
+						if err == nil {
+							ctrl.appRefreshQueue.AddRateLimited(key)
+							ctrl.appOperationQueue.AddRateLimited(key)
```

Comment:
No need to add this, it would be added by refresh processing code after refresh is completed.

Reply:
This resync(enqueue) code is necessary. Otherwise, even if the shard number is changed, the application controller will not process the applications immediately, but will wait for the default resync time.

Reply:
Would refresh not be triggered immediately?

Reply:
With this resync code, it will work, otherwise it will not.

Reply:
The code has changes in 2.13, where refresh enqueues app operation after being done. This prevents some race conditions and makes sure the sync is done with a fresh version of data.

Reply:
So basically we need two resyncs, one initially and one after the refresh?

Reply:
> the application controller will not process the applications immediately, but will wait for the default resync time.

Why would that happen?

Reply:
> So basically we need two resyncs, one initially and one after the refresh?

Yes, I have seen the changes, I will delete this line  `ctrl.appOperationQueue.AddRateLimited(key)`

Reply:
Ah okay. Good to clarify. We have one case where we may enqueue app operation immediately and another after some delay.